More powerful derived kpis

aaron-mcdaid-zalando opened the pull request:
Previously, a derived KPI specified as  `X/Y` would be treated by dividing each element of `X` by the *average* value of `Y`, and then analysis would proceed on these adjusted values:

     adjusted_i = x_i / mean(Y)

This, correctly, meant that the `mean(Adjusted)` was equal the weighted mean of all the `x_i/y_i`. However the variance was overestimated and therefore less powerful. This PR improves the power

The priority in this PR is `make_delta`. This PR doesn't not attempt to apply this fix to `make_group_sequential`, as that will soon be removed and replaced by other changes that are coming. Bootstrapping is also not affected by this PR (but perhaps it should be!)

aaron-mcdaid-zalando committed:
Fix for the possibility that both variances are zero, this will come up in our tests

aaron-mcdaid-zalando committed:
small fix to help when the pool variance is zero

aaron-mcdaid-zalando committed:
A new function (to be used in the next commit) 'normal_sample_weighted_difference' to do derived kpis correctly

aaron-mcdaid-zalando committed:
extend 'delta' to take the weights too, and to call 'normal_sample_weighted_difference'. The p-value will still be wrong though

aaron-mcdaid-zalando committed:
Compute the p-value in 'normal_sample_weighted_difference' also

aaron-mcdaid-zalando committed:
'_get_denominators'

aaron-mcdaid-zalando committed:
'delta' in 'expan/core/statistics.py' now takes the *_denominators args also

aaron-mcdaid-zalando committed:
still replicating the old (wrong) behaviour, just rewriting the code more understandably

aaron-mcdaid-zalando committed:
The closures in 'make_group_sequential', 'make_bayes_factor', and 'make_bayes_precision' now accept the two extra args, but they simply divide by the mean. This continues the old behaviour. Ideally, we would update all three of them

aaron-mcdaid-zalando committed:
Defined '_{x,y}_strange to keep the same (unpowerful) behaviour as before

aaron-mcdaid-zalando committed:
Finally, passing in the numerators and denominators as we should

aaron-mcdaid-zalando committed:
A few tests for the derived_kpis changes

aaron-mcdaid-zalando committed:
s/np.mean/np.nanmean/

aaron-mcdaid-zalando commented:
OK, I'm happier now with this version of the code. Let's start reviewing!

I'm going to write another test though, to quickly run a large number of A/A tests (with small sample size) to confirm that the p-values are nicely distributed

daryadedik commented on expan/core/statistics.py:
```diff
@@ -85,16 +110,28 @@ def delta(x, y, assume_normal=True, alpha=0.05, min_observations=20, nruns=10000
             logger.info("The distribution of two samples is assumed normal. "
                         "Performing the sample difference distribution calculation.")
-            c_i = normal_sample_difference(x=_x, y=_y, percentiles=percentiles, relative=relative)
+            lots_of_info = normal_sample_weighted_difference(x_numerators=_x, y_numerators=_y, x_denominators=_x_denominators, y_denominators = _y_denominators, percentiles=percentiles, relative=relative)
```
this name of the variable is not straightforward to understand. Maybe some more meaningful name?

aaron-mcdaid-zalando replied:
Good point :-)  It's quite similar to the delta_statistics. I might just call it `partial_delta_statistics`. I'll push something after my next meeting

daryadedik commented on expan/core/early_stopping.py:
```diff
@@ -38,7 +38,14 @@ def obrien_fleming(information_fraction, alpha=0.05):
+
+        # these next too lines are wrong, but they are bug-compatible with v0.6.13 !
+        x = x / np.nanmean(x_denominators)
+        y = y / np.nanmean(y_denominators)
```
@aaron-mcdaid-zalando question for my understanding: why we need that devisions?

aaron-mcdaid-zalando replied:
The current system, the "weighting trick", simply divides each numerator by the mean of the denominators. That is what I am replicating here.

For `fixed_horizon`, I'm doing the right thing. For the other three (including this one) I'm simply replicating the existing behaviour.

The big picture is that the interface to a worker has been changed:
  - for a direct kpi, pass the two samples as-is
  - for derived kpis, pass four samples, i.e. the numerators and denominators for control and also for treatment

aaron-mcdaid-zalando committed:
tests/test_derived.py: same tests, just rewritten much more compactly

aaron-mcdaid-zalando committed:
rename 'lots_of_info' to 'partial_simple_test_stats'

aaron-mcdaid-zalando committed:
'test_using_lots_of_AA_tests': a more thorough test that the p-values are uniform

aaron-mcdaid-zalando committed:
Now pass 'ddof=1' to np.var and np.std within 'fixed_horizon'. This is to ensure our pooled_std is correct. For large samples, this change is negligible

aaron-mcdaid-zalando reviewed (commented):
This change (adding `ddof=1` to some calls) makes a small numerical change, but it's needed to ensure our pooled_variance estimate is correct. This change basically ensures the variance estimate is divided by (n-1), not (n) as is the default in numpy.

See the defininition of `s_i^2` in https://en.wikipedia.org/wiki/Pooled_variance#Definition_and_computation

As a result, I had to make small changes to some tests in this commit also

aaron-mcdaid-zalando committed:
'normal_sample_difference' now simply repackages the input and forwards it to 'normal_sample_weighted_difference'. A small tweak to test results is required here also, as we are now using the correct 'ddof' in np.var and np.std

shansfolder commented:
@aaron-mcdaid-zalando thanks for this great fix. I'll take a closer look later today, but I have three general comments atm:

1. I merged your other PR, and now this PR has merge conflict with ```master``` branch you might want to resolve.
2. I've given admin to ```aaron-mcdaid-zalando```. You should be able to merge  now.
3. Please also update corresponding documentation in the ```/docs``` folder when this PR is finished.

Thanks a lot again!

aaron-mcdaid-zalando committed:
Merge branch 'master' into more_powerful_derived_kpis

aaron-mcdaid-zalando committed:
Update API doc for 'normal_sample_weighted_difference'

aaron-mcdaid-zalando committed:
Update the glossary with a detailed definition of the new formula for the variance of derived KPIs

aaron-mcdaid-zalando commented:
I updated documents in a two commits this afternoon

aaron-mcdaid-zalando committed:
Derived KPIs: better handling of zeros and Nans, to be more consistent with older versions of expan.

aaron-mcdaid-zalando committed:
Better test, 'test_using_lots_of_AA_tests_ratio__always_1__withZeros', to really test what happens with division by zero

aaron-mcdaid-zalando committed:
Make sure all NaNs have been removed (they are removed just a few lines above)

aaron-mcdaid-zalando commented:
This morning, Monday, I pushed another few important commits to this. If a derived KPI is `5/0` then we should remove it. This might seem crazy, but we sometimes see that in real data. My first commits in this PR were a bit inconsistent about how they handled zeros and NaNs

Simply put: Any datapoint with NaN in the numerator should be discarded. Also, any datapoint with zero or NaN in the denominator should be discarded.

aaron-mcdaid-zalando committed:
Merge branch 'master' (1.2.5) into more_powerful_derived_kpis

aaron-mcdaid-zalando committed:
bugfix: that code didn't handle pandas Series very well. Now, we convert to np.array(dtype=float) as soon as possible

aaron-mcdaid-zalando committed:
No need for that coercion any more, they are already of the right type

aaron-mcdaid-zalando committed:
Four variables no longer used in this function, let's delete them for simplicity

aaron-mcdaid-zalando commented:
I merged with master (v1.2.5) this morning, and then discovered a bug. So I pushed a fix.

Earlier in this PR, I rewrote `delta` to handle the derived KPIs in a more correct way. But, while doing that, I broke support for pandas Series in the function. Now, I've changed this function so that all inputs are changed to `np.array(dtype=float)` as soon as possible

Tests passing again now

gbordyugov commented:
My only wish would be to have a little intro to the documentation section explaining the problem this PR is solving.

aaron-mcdaid-zalando committed:
Delete '_get_weights' and '{control,treatment}_data', as they are not used any more

aaron-mcdaid-zalando committed:
Update the Glossary to justify the new formula a little more

aaron-mcdaid-zalando committed:
'tests/test_derived.py': stop hardcoding the '1000'. Hopefully this test is now easier to understand

aaron-mcdaid-zalando committed:
A test with a little power. I.e. add a small treatment effect and confirm the pvalues are pulled down to zero

aaron-mcdaid-zalando commented:
A few changes pushed since the last review:

- A little extra commentary in the Glossary
- delete some variables that are computed, but no longer used. And delete the `_get_weights` function as it's no longer used
- Improve the testing a little, in particular adding a test where there is a non-zero effect. This is done to confirm that the pvalues are pulled down

aaron-mcdaid-zalando committed:
Make a 'BETA' variable more explicit

gbordyugov commented on expan/core/statistics.py:
```diff
@@ -16,10 +18,10 @@ def _delta_mean(x, y):
 def make_delta(assume_normal=True, alpha=0.05, min_observations=20, nruns=10000, relative=False):
     """ A closure to the delta function. """
-    return lambda x, y: delta(x, y, assume_normal, alpha, min_observations, nruns, relative)
+    return lambda x, y, x_denominators=1, y_denominators=1: delta(x, y, x_denominators, y_denominators, assume_normal, alpha, min_observations, nruns, relative)
```
I think that writing out this lambda with `def` would make the code a bit more readable

aaron-mcdaid-zalando replied:
Done. A local `def` named `go`

gbordyugov commented:
I think we can merge this

aaron-mcdaid-zalando committed:
In 'make_delta', rewrite the lambda as a local def (called 'go')
